src/attacks.rs

```diff
@@ -14,16 +20,171 @@ pub enum DepthReduceSet {
+#[derive(Clone, Default)]
+pub struct SingleAttackResult {
+    depth: f64,
+    exclusion_size: f64,
+    // graph_size: usize,
+    // FIXME: Do we care to know the absolute number or just
+    // relative to the graph size?
+}
+
+impl std::fmt::Display for SingleAttackResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "\t-> |S| = {:.4}\n\t-> depth(G-S) = {:.4}",
+            self.exclusion_size, self.depth,
+        )
+    }
+}
+
+pub fn attack(g: &mut Graph, attack: DepthReduceSet) -> SingleAttackResult {
+    let start = Instant::now();
+    let set = depth_reduce(g, attack);
+    let duration = start.elapsed();
+    let depth = g.depth_exclude(&set);
+    let result = SingleAttackResult {
+        depth: depth as f64 / g.size() as f64,
+        exclusion_size: set.len() as f64 / g.size() as f64,
+    };
+    println!("{}", result);
+    println!("\t-> time elapsed: {:?}", duration);
+    result
+}
+
+// FIXME: Eventually this should replace the old `attack`.
+pub fn attack_with_profile(g: &mut Graph, profile: AttackProfile) {
+    let mut targets: Vec<f64> = Vec::new();
+    let mut target = profile.range.start;
+    loop {
+        targets.push(target);
+        target += profile.range.interval;
+        if target >= profile.range.end {
+            break;
+        }
     }
+    // FIXME: Move this logic to `TargetRange`.
+
+    let mut results: Vec<Vec<SingleAttackResult>> =
+        vec![vec![SingleAttackResult::default(); profile.runs]; targets.len()];
+
+    for (t, target) in targets.iter().enumerate() {
+        let absolute_target = (target * g.size() as f64) as usize;
+        let attack_type = match profile.attack.clone() {
+            DepthReduceSet::ValiantDepth(_) => DepthReduceSet::ValiantDepth(absolute_target),
+            DepthReduceSet::ValiantSize(_) => DepthReduceSet::ValiantSize(absolute_target),
+            DepthReduceSet::ValiantAB16(_) => DepthReduceSet::ValiantAB16(absolute_target),
+            DepthReduceSet::GreedyDepth(_, p) => DepthReduceSet::GreedyDepth(absolute_target, p),
+            DepthReduceSet::GreedySize(_, p) => DepthReduceSet::GreedySize(absolute_target, p),
+        };
+        // FIXME: Same as before, the target should be decoupled from the type of attack.
+
+        for run in 0..profile.runs {
+            // FIXME: Should we create a new graph for different runs? Or targets?
+
+            println!(
+                "Attack (run {}) target ({:?} = {}), with {:?}",
+                run, profile.target, target, attack_type
+            );
+            results[t][run] = attack(g, attack_type.clone());
+        }
+    }
+
+    // FIXME: Define how to process and output results, e.g., average across runs
+    //  and take best result across different targets. Should we then output here
+    //  a single result or an entire array or a combination (best + all)?
```

Comment:
^

Reply:
We should output all data and then we later filter that according to our needs, outside this Rust code.

Reply:
Understood, could you be more specific in terms of the format and layout needed?